run.c

```diff
@@ -421,16 +421,49 @@ float random_f32() { // random float32 in [0,1)
     return (random_u32() >> 8) / 16777216.0f;
 }
 
-int sample(float* probabilities, int n) {
-    // sample index from probabilities, they must sum to 1
-    float r = random_f32();
-    float cdf = 0.0f;
+typedef struct {
+    float value;
+    int index;
+} IndexedFloat;
+
+int compare_indexed_float(const void* a, const void* b) {
+    IndexedFloat* indexed_a = (IndexedFloat*)a;
+    IndexedFloat* indexed_b = (IndexedFloat*)b;
+    // sorting in descending order
+    return (indexed_b->value > indexed_a->value) ? 1 : ((indexed_b->value < indexed_a->value) ? -1 : 0);
+}
+
+int sample_top_p(float* probs, int n, float topp) {
+    IndexedFloat probs_sort[n];
     for (int i = 0; i < n; i++) {
-        cdf += probabilities[i];
-        if (r < cdf) {
-            return i;
+        probs_sort[i].value = probs[i];
+        probs_sort[i].index = i;
+    }
+
+    qsort(probs_sort, n, sizeof(IndexedFloat), compare_indexed_float);
+
+    // Calculate cumulative probabilities
+    float cum_probs[n];
+    cum_probs[0] = probs_sort[0].value;
+    int p = 1;
+    for (; cum_probs[p-1]<=topp && p < n; p++) {
+        cum_probs[p] = cum_probs[p - 1] + probs_sort[p].value;
+    }
+
+    for (int i = 0; i < p; i++) {
+        probs_sort[i].value /= cum_probs[p-1];
+    }
+
+    // Sample next token based on the modified probability distribution
+    float r = random_f32();
+    float cum_prob = 0.0;
+    for (int i = 0; i < p; i++) {
+        cum_prob += probs_sort[i].value;
+        if (cum_prob > r) {
+            return probs_sort[i].index;
         }
     }
+
     return n - 1; // in case of rounding errors
```

Comment:
Not sure if it makes a difference, but shouldn't this be `p-1`? Because at this point, we already know that the sum up until the `p-th` element is larger than `topp` and therefore `p-1` should be that last valid index to return.

Reply:
Suppose we have the following probability distribution and top-p threshold:

    probs = [0.1, 0.2, 0.3, 0.4]
    topp = 0.7

We want to sample an index from the top 70% of probabilities.
Working Through the Code

    Sort and Accumulate: The code will sort the probabilities and accumulate them until the accumulated value exceeds topp. In this case:
        probs_sort = [(0.4, 3), (0.3, 2), (0.2, 1), (0.1, 0)]
        The accumulation will stop at p = 3 since 0.4+0.3+0.2=0.9>0.70.4+0.3+0.2=0.9>0.7.

    Sampling Loop: The loop will consider probabilities from index 0 to p. In this example, it will consider indices 0, 1, and 2.

    Rounding Errors: If, due to rounding errors, no index is returned from the loop, we reach the last line of the function.

Return Value

    Returning p−1p−1: This would return the index 2, which is part of the top 70% of probabilities.
    Returning n−1n−1: This would return the index 3, which is also part of the top 70% of probabilities.

Counter Example

However, consider a slight change in the probabilities:

    probs = [0.4, 0.3, 0.2, 0.1]
    topp = 0.7

In this case:

    probs_sort = [(0.4, 0), (0.3, 1), (0.2, 2), (0.1, 3)]
    The accumulation will again stop at p = 3.

But now:

    Returning p−1p−1: This would return the index 2, part of the top 70%.
    Returning n−1n−1: This would return the index 3, which is not part of the top 70% of probabilities.

Conclusion

The second example demonstrates a situation where returning n−1n−1 would fail to align with the top-p threshold, as it returns an index not part of the top-p probabilities. By returning p−1p−1, the code would ensure that the return value is consistent with the top-p threshold, even in this specific case.

Thank you LLMs for that succinct appraisal, what tring said.

Reply:
Thank you both for advice. fixed